store/postgres/src/entities.rs

```diff
@@ -894,6 +907,51 @@ fn entity_from_json(json: serde_json::Value, entity: &str) -> Result<Entity, Sto
     Ok(value)
 }
 
+/// Determine the `IdType` for a subgraph
+// Only public so that tests can access it
+pub fn id_type(conn: &PgConnection, subgraph: &SubgraphDeploymentId) -> Result<IdType, StoreError> {
+    use self::public::columns;
+    use self::public::deployment_schemas as ds;
+    use self::public::DeploymentSchemaVersion;
+
+    if subgraph.is_meta() {
+        return Ok(IdType::String);
+    }
+
+    // These two queries should really be one, but it seems impossible to
+    // select from two tables with diesel when doing an inner_join.
+    let (schema_name, version): (String, self::public::DeploymentSchemaVersion) = ds::table
+        .select((ds::name, ds::version))
+        .filter(ds::subgraph.eq(subgraph.as_str()))
+        .first(conn)?;
+    if version == DeploymentSchemaVersion::Split {
+        // JSONB storage always uses `String` as `ID`, but we can't tell
+        // that from the information_schema reliably, because `entity_history`
+        // has an integer `id` column
+        return Ok(IdType::String);
+    }
+
+    let (data_type, table_name): (String, String) = columns::table
+        .select((columns::data_type, columns::table_name))
+        .filter(columns::column_name.eq("id"))
+        .filter(columns::table_schema.eq(schema_name.as_str()))
+        .first(conn)?;
```

Comment:
Ok, I can see how this could easily be changed to obtain the ID type per entity.